lib/vast/Tower/Link.cpp

```diff
@@ -0,0 +1,171 @@
+
+    op_mapping reverse_mapping(const op_mapping &from) {
+        op_mapping out;
+        for (const auto &[root, ops] : from) {
+            for (auto op : ops) {
+                out[op].push_back(root);
+            }
+        }
+        return out;
+    }
+
+    void dbg(const op_mapping &mapping, auto &outs) {
+        outs << "Mapping:\n";
+
+        auto flag = mlir::OpPrintingFlags().skipRegions();
+        auto render_op = [&](operation op) -> decltype(outs) & {
+            outs << op << ": ";
+            op->print(outs, flag);
+            if (!op->getRegions().empty())
+                outs << " ... regions ...";
+            return outs;
+        };
+
+        for (const auto &[from, to] : mapping) {
+            outs << "== ";
+            render_op(from) << "\n";
+            for (auto t : to) {
+                outs << "   -> ";
+                render_op(t) << "\n";
+            }
+        }
+    }
+
+    using loc_to_op_t = llvm::DenseMap< loc_t, operations >;
+
+    void dbg(const loc_to_op_t &mapping, auto &outs) {
+        outs << "loc_to_op\n";
+        for (const auto &[loc, to] : mapping) {
+            outs << ".." << loc << "\n";
+            for (auto t : to) {
+                outs << "   -> " << *t << "\n";
+            }
+        }
+    }
+
+    loc_to_op_t gather_map(location_info_t &li, operation op) {
+        loc_to_op_t out;
+        auto collect = [&](operation op) { out[li.self(op)].push_back(op); };
+        op->walk(collect);
+        return out;
+    }
+
+    std::vector< op_mapping > mk_link_mappings(const auto &links) {
+        std::vector< op_mapping > out;
+        for (const auto &l : links)
+            out.emplace_back(l->parents_to_children());
+        return out;
+    }
+
+    // `transition` handles the lookup between levels - we need this as we want to generalise
+    // (`levels` can be arbitrary mapping, not just `op -> { op }`).
+    op_mapping build_map(op_mapping init, const auto &levels, auto transition) {
+        auto handle_level = [&](const auto &level) {
+            auto handle_element = [&](operation op) -> operations {
+                auto prev_ops = level.find(transition(op));
+                if (prev_ops == level.end())
+                    return {};
+                return prev_ops->second;
+            };
+
+            for (auto &[op, todo] : init) {
+                operations parents;
+                for (auto current : todo) {
+                    parents.append_range(handle_element(current));
+                }
+                todo = std::move(parents);
+            }
+        };
+
+        for (const auto &level : levels | std::views::drop(1))
+            handle_level(level);
+
+        return init;
+    }
+
+    op_mapping build_map(const link_vector &links) {
+        auto transition = [](operation op) { return op; };
+        auto init = links.front()->parents_to_children();
+        return build_map(init, mk_link_mappings(links), transition);
```

Comment:
copy of `init` or fix `build_map` to take rvalue?

Reply:
Missing `std::move`, will fix. No need to modify the prototype, the behaviour that there is a copy on call is fine, because the object will be modified as part of algorithm we use.